_posts/2017-06-09-upgrade-to-rails-3.markdown

```diff
@@ -0,0 +1,157 @@
+
+<!--more-->
+
+1. Considerations
+2. Ruby version
+3. Tools
+4. XSS protection
+5. Config files
+6. Gems
+7. Deprecations
+  - Active Record
+  - Action Mailer
+  - Metal
+  - Railties
+8. Next steps
+
+## 1. Considerations
+Before beginning with the upgrade process, we recommend that each version of your Rails app has the latest [patch version](http://semver.org) before moving to the next major/minor version. For example, in order to follow this article, your [Rails version](https://rubygems.org/gems/rails/versions) should be at 2.3.18 before updating to Rails 3.0.20
+
+## 2. Ruby version
+Rails 3.0 requires Ruby 1.8.7 or higher, but no more than 1.9.3. If you want to use Ruby 1.9.x, we recommend you skip directly to 1.9.3. Also Ruby 1.9.1 is not usable because it has segmentation faults on Rails 3.0. That means that the compatible Ruby versions for Rails 3.0 are 1.8.7, 1.9.2, or 1.9.3.
+
+## 3. Tools
+There is an [official plugin](https://github.com/rails/rails_upgrade) that helps the upgrade process. You just need to install the script by doing `script/plugin install git://github.com/rails/rails_upgrade.git` and then run `rake rails:upgrade:check` to see most of the files you need to upgrade in your application. It also provides some other generators to upgrade specific areas in you app like routes or gems.
+
+Sometimes it's also useful to check which files changed between two specifics versions of Rails. Fortunately [this website](http://railsdiff.org/2.3.18/3.0.0) makes that easy.
+
+## 4. XSS protection
+In this version, Rails automatically adds [XSS protection](http://yehudakatz.com/2010/02/01/safebuffers-and-rails-3-0/) in order to scape any content, so you will probably need to update your templates according to this. Luckely, there is an [official plugin](https://github.com/rails/rails_xss) exactly for this purpose. We recommend you to take a look of this.
+
+## 5. Config files
+Rails 3 introduces the concept of an Application object. An application object holds all the specific application configurations and it's similar to the current config/environment.rb from Rails 2.3. The application object is defined in config/application.rb. You should move there most of the configuration that you had in config/environment.rb.
+
+In terms of routes, there are a couple of changes that you need to apply to your routes.rb file. For example:
+
+```
+# Rails 2.3 way:
+ActionController::Routing::Routes.draw do |map|
+  map.resources :products
+end
+
+# Rails 3.0 way:
+AppName::Application.routes do
+  resources :products
+end
+```
+You can go to [this article](https://blog.engineyard.com/2010/the-lowdown-on-routes-in-rails-3) to see in depth about this topic.
+
+## 6. Gems
+[Bundler](https://bundler.io/) is the default way to manage Gem dependencies in Rails 3 applications. You will need to add a [Gemfile](https://bundler.io/v1.15/gemfile_man.html) in the root of your app, define all you gems there, and then get rid of the config.gem.
```

Comment:
@lubc Could you add a couple before/after examples about "getting rid of the config.gem" statements?